### ISSUE TITLE
SegFormer Training extremely slow

### ISSUE THREAD

**Issue by omarequalmars** [qubvel-org/segmentation_models.pytorch#996]
I've been training a series of models implemented by this package for a while, all using 'tu-mobilevit_xxs' as an encoder. However I noticed that the latest addition, segformer, is extremely slow in training compared to the others despite being reported to have the same number of parameters by Pytorch lightning. Here is a visualization from Tensorboard:

![image](https://github.com/user-attachments/assets/25761da5-80a5-41c0-a39c-4f23b3c1d0ae)

And here is the parameter count reported by Lightning:

![image](https://github.com/user-attachments/assets/2324e124-4967-4751-81f5-1495c630b4f3)

It's much slower than the rest in training, and considerably slows down my laptop. Why does it take so much resources despite being approximately the same size as the other models? Here are my architecture hyperparameters for reference, for every model I trained:

```
    h_params_Unet = {
        'encoder': encoder,
        'depth': 5,
        'Bnorm': True,
        'Attn': 'scse',
        'channels': (256, 128, 64, 32, 16),
        'lr_init': 1e-3,
        'weight_decay': 0,
        'temperature': 2,
        'Arch': "Unet"
    }

    h_params_DLV3plus = {
        'encoder': encoder,
        'depth': 5,
        'encoder_output_stride': 16,
        'channels': 512,
        'decoder_atrous_rates': (16, 32, 128),
        'lr_init': 1e-3,
        'weight_decay': 0,
        'temperature': 2,
        'Arch': "DLV3+"
    }

    h_params_MANet = {
        'encoder': encoder,               # Backbone encoder
        '
```depth': 5,                       # Number of encoder stages
        'Bnorm': True,                    # Use batch normalization in the decoder
        'channels': (256, 128, 64, 32, 16),  # Number of channels in each decoder layer
        'decoder_pab_channels': 64,
        'lr_init': 1e-3,                  # Initial learning rate
        'weight_decay': 0,                # Weight decay for regularization
        'temperature': 2,                 # Temperature for distillation
        'Arch': "MANet"
    }

    h_params_PAN = {
        'encoder': encoder,               # Backbone encoder
        'encoder_output_stride': 16,
        'channels': 512,                  # Number of channels in each decoder layer
        'lr_init': 1e-3,                  # Initial learning rate
        'weight_decay': 0,                # Weight decay for regularization
        'temperature': 2,                 # Temperature for distillation
        'Arch': "PAN"
    }

    h_params_Segformer = {
    'encoder': encoder,               # Backbone encoder
    'depth': 5,                       # Number of encoder stages
    'channels': 512,  # Number of channels in each decoder layer
    'lr_init': 1e-3,                  # Initial learning rate
    'weight_decay': 0,                # Weight decay for regularization
    'temperature': 2,                 # Temperature for distillation
    'Arch': "Segformer"
}


All of them have the same number of channels and depth where applicable. What am I doing wrong?

**Comment by qubvel** [comment]
Hey @omarequalmars, thanks for reporting the issue! It might be due to it's high-resolution features in the head.. but there might be some profiling needed to find the reason.

cc @brianhou0208 maybe you have some insights

**Comment by brianhou0208** [comment]
Hi @omarequalmars ,

I tested your hyperparameter settings and resolved performance issue. Please check #998 

Given that you're using a lightweight backbone encoder, I recommend reducing the `decoder_channels` to further optimize inference speed, as highlighted in the [original paper](https://arxiv.org/pdf/2105.15203).

Regarding input resolution, since it wasn't specified, smaller resolutions are unlikely to cause significant speed differences. However, larger resolutions could have a notable impact on performance.

If you're exploring transformer-style semantic segmentation models, aside from SegFormer, you may also consider [TopFormer](https://arxiv.org/abs/2204.05525) and [SeaFormer](https://arxiv.org/abs/2301.13156), which are designed for efficiency as well.

**Comment by omarequalmars** [comment]
Hi @brianhou0208 ,

I'm trying it out now with the decoder channels reduced; running almost as fast as the others. I'm using 512 by 512 image sizes and there are no noticeable slow downs compared to other models trained with identical images sizes. I will definitely be looking into TopFormer and SeaFormer for a comparison. 

Thank you so much for the quick fix, you're a lifesaver!
